### PR TITLE
[KYUUBI #5642][AUTHZ] Support path privilege of CreateDatabaseCommand/AlterDatabaseSetLocationCommand/CreateNamespace/SetNamespaceLocation 

### LINKED ISSUE #5642: [Subtask] Support path privilege of CreateDatabaseCommand AlterDatabaseSetLocationCommand
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/kyuubi/issues) and found no similar issues.

### Describe the subtask

Support path privilege of CreateDatabaseCommand AlterDatabaseSetLocationCommand

### Parent issue

#5368

### Are you willing to submit PR?

- [X] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to improve.
- [ ] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### _Why are the changes needed?_
To close #5642 
Support path privilege of CreateDatabaseCommand/AlterDatabaseSetLocationCommand/CreateNamespace/SetNamespaceLocation

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

### _Was this patch authored or co-authored using generative AI tooling?_
No

### CODE REVIEW

File: extensions/spark/kyuubi-spark-authz/src/main/resources/database_command_spec.json [hunk 1]
```diff
@@ -122,7 +137,13 @@
     "catalogDesc" : null,
     "isInput" : false
   } ],
-  "opType" : "CREATEDATABASE"
+  "opType" : "CREATEDATABASE",
+  "uriDescs" : [ {
+    "fieldName" : "path",
+    "fieldExtractor" : "StringURIExtractor",
+    "actionTypeDesc" : null,
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/resources/database_command_spec.json:144]: we can remove this actionTypeDesc from desc as it's useless. Let's create a separate PR?

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/uriExtractors.scala [hunk 2]
```diff
@@ -33,7 +33,11 @@ object URIExtractor {
  */
 class StringURIExtractor extends URIExtractor {
   override def apply(v1: AnyRef): Seq[Uri] = {
-    Seq(Uri(v1.asInstanceOf[String]))
+    v1 match {
+      case str: String => Seq(Uri(str))
+      case Some(str: String) => Seq(Uri(str))
```
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/uriExtractors.scala:38]:
```suggestion
      case uriPath: String => Seq(Uri(uriPath))
      case Some(uriPath: String) => Seq(Uri(uriPath))
```
Avoid using the data type name as the variable name.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/CommandSpec.scala [hunk 3]
```diff
@@ -57,7 +57,8 @@ trait CommandSpecs[T <: CommandSpec] {
 case class DatabaseCommandSpec(
     classname: String,
     databaseDescs: Seq[DatabaseDesc],
-    opType: String = "QUERY") extends CommandSpec {}
+    opType: String = "QUERY",
+    uriDescs: Seq[UriDesc] = Nil) extends CommandSpec {}
```
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/CommandSpec.scala:61]:
```suggestion
    uriDescs: Seq[UriDesc] = Nil,
    opType: String = "QUERY") extends CommandSpec {}
```
Shall we keep the order in `descs` for different kinds and followed by the `opType`?

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala [hunk 4]
```diff
@@ -1155,4 +1156,25 @@ class HiveCatalogRangerSparkExtensionSuite extends RangerSparkExtensionSuite {
       }
     }
   }
+
+  test("CreateDatabaseCommand/AlterDatabaseSetLocationCommand") {
+    assume(SPARK_RUNTIME_VERSION <= "3.2")
```
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala:1161]:
```suggestion
    assume(!isSparkV33OrGreater)
```

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/uriExtractors.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/uriExtractors.scala]: Done

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala (hunk 4 above)
Comment (AngersZhuuuu) [reply extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala]: Done

File: extensions/spark/kyuubi-spark-authz/src/main/resources/database_command_spec.json (hunk 1 above)
Comment (AngersZhuuuu) [reply extensions/spark/kyuubi-spark-authz/src/main/resources/database_command_spec.json]: DOne

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/CommandSpec.scala (hunk 3 above)
Comment (AngersZhuuuu) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/CommandSpec.scala]: Hmmm if we want to do this, better in a separated pr and do this for all CommandSpec

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala [hunk 5]
```diff
@@ -19,6 +19,7 @@ package org.apache.kyuubi.plugin.spark.authz.ranger
 
 import java.nio.file.Path
 
+import scala.collection.Seq
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala:22]:
```suggestion
```

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala [hunk 6]
```diff
@@ -1155,4 +1156,25 @@ class HiveCatalogRangerSparkExtensionSuite extends RangerSparkExtensionSuite {
       }
     }
   }
+
+  test("CreateDatabaseCommand/AlterDatabaseSetLocationCommand") {
+    assume(!isSparkV33OrGreater)
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala:1161]: Let's support 3.5

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala (hunk 5 above)
Comment (AngersZhuuuu) [reply extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala]: Done

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala (hunk 6 above)
Comment (AngersZhuuuu) [reply extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala]: Also support CreateNamespace and AlterNamespaceSetLocation command in this pr, then we don't need this.

**Comment by AngersZhuuuu** [comment]
ping @yaooqinn @bowenliang123

**Comment by bowenliang123** [comment]
Thanks for extending the privilege checks for path base commands, which is general and important for lakehouse usages.
LGTM overall.

Could you add some positive test cases to show allowed queries and the neccessary policies for tests?

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/V2CommandsPrivilegesSuite.scala [hunk 7]
```diff
@@ -738,7 +738,7 @@ abstract class V2CommandsPrivilegesSuite extends PrivilegesBuilderSuite {
       "org.apache.spark.sql.catalyst.plans.logical.SetNamespaceLocation")
     assert(operationType === ALTERDATABASE_LOCATION)
     assert(in.isEmpty)
-    assert(out.size === 1)
+    assert(out.size === 2)
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/V2CommandsPrivilegesSuite.scala:741]: Can we assert out.last?

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala [hunk 8]
```diff
@@ -1172,4 +1172,24 @@ class HiveCatalogRangerSparkExtensionSuite extends RangerSparkExtensionSuite {
...
   }
+
+  test("CreateDatabaseCommand/AlterDatabaseSetLocationCommand") {
+    val db1 = "db1"
+    withSingleCallEnabled {
+      withTempDir { path1 =>
+        withTempDir { path2 =>
+          withCleanTmpResources(Seq((s"$db1", "database"))) {
+            interceptContains[AccessControlException](
+              doAs(someone, sql(s"CREATE DATABASE $db1 LOCATION '$path1'")))(
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala:1183]: CREATE DATABASE w/o location clause will not check path privilege, right?

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/uriExtractors.scala [hunk 9]
```diff
@@ -49,12 +53,18 @@ class CatalogStorageFormatURIExtractor extends URIExtractor {
   }
 }
 
-class OptionsUriExtractor extends URIExtractor {
+class PropertiesPathUriExtractor extends URIExtractor {
   override def apply(v1: AnyRef): Seq[Uri] = {
     v1.asInstanceOf[Map[String, String]].get("path").map(Uri).toSeq
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/uriExtractors.scala:58]: This behavior actually differs from spark who will use a case-insensitive map.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/uriExtractors.scala [hunk 10]
```diff
@@ -49,12 +53,18 @@ class CatalogStorageFormatURIExtractor extends URIExtractor {
...
   }
 }
 
+class PropertiesLocationUriExtractor extends URIExtractor {
+  override def apply(v1: AnyRef): Seq[Uri] = {
+    v1.asInstanceOf[Map[String, String]].get("location").map(Uri).toSeq
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/uriExtractors.scala:64]: ditto

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala (hunk 8 above)
Comment (AngersZhuuuu) [reply extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala:1183]: > CREATE DATABASE w/o location clause will not check path privilege, right?

Yea

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/V2CommandsPrivilegesSuite.scala (hunk 7 above)
Comment (AngersZhuuuu) [reply extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/V2CommandsPrivilegesSuite.scala:741]: DOne

**Comment by AngersZhuuuu** [comment]
> Thanks for extending the privilege checks for path base commands, which is general and important for lakehouse usages. LGTM overall.
> 
> Could you add some positive test cases to show allowed queries and the neccessary policies for tests?

Done

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/uriExtractors.scala (hunk 10 above)
Comment (AngersZhuuuu) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/uriExtractors.scala:64]: Spark side code
截屏2023-11-09 上午10 46 15
截屏2023-11-09 上午10 46 56

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/uriExtractors.scala (hunk 9 above)
Comment (AngersZhuuuu) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/uriExtractors.scala:58]: > This behavior actually differs from spark who will use a case-insensitive map.

But spark for such command all use lower case
截屏2023-11-09 上午10 48 11

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
thanks, merged to master

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala [hunk 11]
```diff
@@ -1172,4 +1173,29 @@ class HiveCatalogRangerSparkExtensionSuite extends RangerSparkExtensionSuite {
...
+              s"does not have [create] privilege on [$db1], " +
+                s"[write] privilege on [[$path1, $path1/]]")
+            doAs(admin, sql(s"CREATE DATABASE $db1 LOCATION '$path1'"))
+            interceptContains[AccessControlException](
+              doAs(someone, sql(s"ALTER DATABASE $db1 SET LOCATION '$path2'")))(
+              s"does not have [alter] privilege on [$db1], " +
+                s"[write] privilege on [[$path2, $path2/]]")
```
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala:1191]: There is a case should be also tested , for the user have alter privilege of the table but lack of write privilege of the  location path.

This should reflect the possible impact to the exited user of authz plugin.

@AngersZhuuuu @yaooqinn

**Pull request closed without merging**